proof/refine/ARM/KHeap_R.thy

```diff
@@ -3917,6 +3914,180 @@ lemma refillSufficient_wp:
+  assumes
+      R1: "\<forall>s s'. (s, s') \<in> state_relation \<longrightarrow>
+         P s \<longrightarrow> P' s' \<longrightarrow> sc_at ptr s \<longrightarrow> sc_at' ptr s' \<longrightarrow>
+           (\<forall>n. (((\<lambda>ko. obj_bits ko = min_sched_context_bits + n) |< kheap s) ptr) \<longrightarrow>
+           sc_relation (the ((scs_of2 s ||> f) ptr)) n (the ((scs_of' s' ||> f') ptr)))"
+  and R2: "\<forall>s s'. (s, s') \<in> state_relation \<longrightarrow>
+         P s \<longrightarrow> P' s' \<longrightarrow> sc_at ptr s \<longrightarrow> sc_at' ptr s' \<longrightarrow>
+           heap_ls (replyPrevs_of s')  (scReply (the ((scs_of' s' ||> f') ptr)))
+             (sc_replies (the ((scs_of2 s ||> f) ptr)))"
+  and sz: "\<forall>sc'::sched_context. objBits sc' = objBits (f' sc')"
+  shows
+  "\<lbrakk>(s, s') \<in> state_relation; P s; P' s'; sc_at ptr s; sc_at' ptr s'\<rbrakk> \<Longrightarrow>
+     (kheap_update (\<lambda>hp p. if p = ptr
+                           then
+                             case hp ptr of
+                                Some (kernel_object.SchedContext sc n)
+                                   \<Rightarrow> Some (kernel_object.SchedContext (f sc) n)
+                               | _ \<Rightarrow> hp ptr
+                           else hp p) s,
+     (ksPSpace_update (\<lambda>hp' p. if p = ptr
+                               then case hp' ptr of
+                                  Some (KOSchedContext sc')
+                                     \<Rightarrow> Some (KOSchedContext (f' sc'))
+                                 | _ \<Rightarrow> hp' ptr
+                                else hp' p)) s') \<in> state_relation"
+  supply pred_map_rewrite[simp] scs_of_rewrite[simp] opt_map_left_Some[simp]
+         sc_replies_of_rewrite[simplified, simp]
+  proof -
+  have z': "\<And>s. sc_at' ptr s
+               \<Longrightarrow> \<forall>sc'::sched_context. map_to_ctes ((\<lambda>hp' p. if p = ptr then case hp' ptr of
+                              Some (KOSchedContext sc') \<Rightarrow> Some (KOSchedContext (f' sc'))
+                            | _ \<Rightarrow> hp' ptr else hp' p) (ksPSpace s)) = map_to_ctes (ksPSpace s)"
+    by (clarsimp simp: obj_at_simps fun_upd_def[symmetric])
+  have z: "\<And>s sc'::sched_context. ko_at' sc' ptr s
+               \<Longrightarrow> map_to_ctes (ksPSpace s(ptr \<mapsto> KOSchedContext (f' sc'))) = map_to_ctes (ksPSpace s)"
+    by (clarsimp simp: obj_at_simps)
+  have S: "\<And>(v::'a::pspace_storable). (1 :: word32) < 2 ^ (objBits v)"
+    by (clarsimp simp: obj_at_simps objBits_defs pteBits_def pdeBits_def scBits_pos_power2
+                split: kernel_object.splits arch_kernel_object.splits)
+  assume H: "(s, s') \<in> state_relation" "P s" "P' s'" "sc_at ptr s" "sc_at' ptr s'"
+  show ?thesis
+    using H S sz
+  apply -
+    apply (insert R1[rule_format, OF H(1) H(2) H(3) H(4) H(5)]
+                  R2[rule_format, OF H(1) H(2) H(3) H(4) H(5)])
```

Comment:
tip: You can just say `OF H` if `H` has exactly 5 lemmas in the right order.

Reply:
ah, that’s good to know!